File: Scripts/02_mapa_restaurantes.py
"""
=============================================================
MAPA INTERACTIVO - RESTAURANTES MEDITERRÁNEOS EN BOGOTÁ
Python + Folium  (reemplaza 02_mapa_restaurantes.R)
=============================================================

Estructura del proyecto:
  DoingEconomics_Proyect/
  ├── Data/
  │   └── Raw/
  │       ├── restaurantes_mediterraneos_bogota.csv   <- INPUT
  │       └── Barrios_Bogota/                         <- INPUT shapefile
  └── Outputs/
      └── Graphs/
          └── mapa_restaurantes_mediterraneos.html    <- OUTPUT

Columnas del CSV:
  id | place_id | nombre | direccion | lat | lon |
  rating | num_reviews | google_maps_url | tipos | query_origen | barrio

pip install folium geopandas pandas
=============================================================
"""

import pandas as pd
import geopandas as gpd
import folium
from folium.plugins import MarkerCluster
from pathlib import Path
import html

# ─── RUTAS DEL PROYECTO ───────────────────────────────────────────────────────
SCRIPT_DIR   = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent
DATA_RAW_DIR = PROJECT_ROOT / "Data"    / "Raw"
GRAPHS_DIR   = PROJECT_ROOT / "Outputs" / "Graphs"
BARRIOS_DIR  = DATA_RAW_DIR / "Barrios_Bogota"

GRAPHS_DIR.mkdir(parents=True, exist_ok=True)

INPUT_CSV   = DATA_RAW_DIR / "restaurantes_mediterraneos_bogota.csv"

OUTPUT_HTML = GRAPHS_DIR / f"mapa_restaurantes_mediterraneos.html"

# Nombre del campo de barrio en el shapefile — ajusta si es diferente
CAMPO_BARRIO = "SCANOMBRE"

# ─── CONFIGURACIÓN DE CLUSTERS ────────────────────────────────────────────────
#
#   USE_CLUSTERS = True   →  puntos agrupados en clusters numerados
#                            (útil para ver densidad y navegar rápido)
#
#   USE_CLUSTERS = False  →  todos los puntos visibles simultáneamente
#                            (útil para ver distribución espacial completa)
#
#   El script genera automáticamente un nombre de archivo diferente
#   según la opción, para que puedas tener ambas versiones guardadas.

USE_CLUSTERS = False      # ← cambia aquí: True / False

# ─── PALETA DE COLORES POR COCINA ─────────────────────────────────────────────
COLORES = {
    "Italiana":     "#2A9D8F",
    "Española":     "#E63946",
    "Griega":       "#457B9D",
    "Mediterránea": "#F4A261",
}

def clasificar_cocina(tipos: str, query_origen: str) -> str:
    t = str(tipos).lower()
    q = str(query_origen).lower()
    if "grieg" in t or "greek" in t or "grieg" in q:
        return "Griega"
    if "españ" in t or "spanish" in t or "españ" in q or "tapas" in q:
        return "Española"
    if "italian" in t or "italian" in q or "pizza" in q or "trattoria" in q:
        return "Italiana"
    return "Mediterránea"

# ─── 1. CARGAR CSV ────────────────────────────────────────────────────────────
print("=" * 55)
print("  MAPA INTERACTIVO — RESTAURANTES MEDITERRÁNEOS")
print("=" * 55)

if not INPUT_CSV.exists():
    raise FileNotFoundError(f"CSV no encontrado:\n  {INPUT_CSV}")

df = pd.read_csv(INPUT_CSV, encoding="utf-8-sig")
df = df.dropna(subset=["lat", "lon"]).copy()
df["id"]          = df["id"].apply(lambda i: f"{int(i):03d}")
df["rating"]      = pd.to_numeric(df["rating"],      errors="coerce").round(1)
df["num_reviews"] = pd.to_numeric(df["num_reviews"], errors="coerce").fillna(0).astype(int)
df["barrio"]      = df["barrio"].fillna("Sin datos").replace("", "Sin datos")
df["tipos"]       = df["tipos"].fillna("").astype(str)
df["query_origen"]= df["query_origen"].fillna("").astype(str)
df["cocina"]      = df.apply(
    lambda r: clasificar_cocina(r["tipos"], r["query_origen"]), axis=1
)

print(f"\n✅ Restaurantes cargados : {len(df)}")
print(f"   Distribución por cocina:")
for c, n in df["cocina"].value_counts().items():
    print(f"     {c:<15} : {n}")

# ─── 2. CARGAR SHAPEFILE DE BARRIOS ───────────────────────────────────────────
shp_files = list(BARRIOS_DIR.glob("*.shp"))
if not shp_files:
    raise FileNotFoundError(
        f"No se encontró ningún .shp en:\n  {BARRIOS_DIR}"
    )

barrios = gpd.read_file(shp_files[0]).to_crs(epsg=4326)
print(f"\n✅ Shapefile cargado     : {shp_files[0].name}")
print(f"   Polígonos             : {len(barrios)}")
print(f"   Columnas              : {list(barrios.columns)}")

if CAMPO_BARRIO not in barrios.columns:
    raise KeyError(
        f"\nCampo '{CAMPO_BARRIO}' no existe en el shapefile.\n"
        f"Columnas disponibles: {list(barrios.columns)}\n"
        f"Ajusta CAMPO_BARRIO en la línea 53."
    )

# ─── 3. CONSTRUIR MAPA BASE ───────────────────────────────────────────────────
mapa = folium.Map(
    location=[4.7110, -74.0721],
    zoom_start=11,
    tiles=None                    # tiles se agregan como capas intercambiables
)

# Tiles base
folium.TileLayer(
    tiles="CartoDB positron",
    name="Claro",
    control=True
).add_to(mapa)

folium.TileLayer(
    tiles="OpenStreetMap",
    name="Detallado",
    control=True
).add_to(mapa)

# ─── 4. CAPA DE BARRIOS ───────────────────────────────────────────────────────
barrios_layer = folium.FeatureGroup(name="Barrios", show=True)

folium.GeoJson(
    data=barrios.__geo_interface__,
    style_function=lambda _: {
        "fillColor":   "#f0f0f0",
        "color":       "#999999",
        "weight":       0.8,
        "fillOpacity":  0.2,
    },
    highlight_function=lambda _: {},   # sin resaltado al hover
    interactive=False,                  # bloquea clic y tooltip
).add_to(barrios_layer)

barrios_layer.add_to(mapa)

# ─── 5. FUNCIÓN DE POPUP ──────────────────────────────────────────────────────
def hacer_popup(row: pd.Series) -> str:
    """
    Genera el HTML del popup para cada restaurante.
    html.escape() protege contra caracteres especiales en los datos
    que podrían romper el HTML (&, <, >, comillas).
    """
    def esc(v):
        return html.escape(str(v)) if pd.notna(v) else "-"

    nombre    = esc(row["nombre"])
    rid       = esc(row["id"])
    cocina    = esc(row["cocina"])
    barrio    = esc(row["barrio"])
    direccion = esc(row["direccion"])
    rating    = f"{row['rating']} / 5" if pd.notna(row["rating"]) else "Sin datos"
    reviews   = f"{row['num_reviews']:,}".replace(",", ".") if row["num_reviews"] > 0 else "Sin datos"
    tipos     = esc(row["tipos"]).replace("_", " ") if row["tipos"] else "Sin datos"

    maps_link = (
        f'<a href="{html.escape(row["google_maps_url"])}" target="_blank">'
        f'Ver en Google Maps</a>'
        if pd.notna(row.get("google_maps_url")) and row["google_maps_url"] != ""
        else ""
    )

    return f"""
    <div style="font-family:Arial,sans-serif; font-size:13px;
                line-height:1.8; min-width:210px;">
        <b style="font-size:14px;">{nombre}</b><br>
        ID: {rid}<br>
        Cocina: {cocina}<br>
        Barrio: {barrio}<br>
        Dirección: {direccion}<br>
        Rating: {rating}<br>
        Reseñas: {reviews}<br>
        Tipo: {tipos}<br>
        {maps_link}
    </div>
    """

# ─── 6. CAPA DE RESTAURANTES — CLUSTER O PUNTOS INDIVIDUALES ─────────────────
restaurantes_layer = folium.FeatureGroup(name="Restaurantes", show=True)

if USE_CLUSTERS:
    # ── Modo cluster: agrupa puntos cercanos en burbujas numeradas ────────────
    print("\n🔵 Modo: CLUSTERS activado")
    contenedor = MarkerCluster(
        options={
            "showCoverageOnHover": False,
            "zoomToBoundsOnClick": True,
            "spiderfyOnMaxZoom":   True,
        }
    )
    contenedor.add_to(restaurantes_layer)

else:
    # ── Modo puntos: todos los puntos visibles al mismo tiempo ────────────────
    print("\n🟢 Modo: PUNTOS INDIVIDUALES activado")
    contenedor = restaurantes_layer   # los markers van directo a la capa

for _, row in df.iterrows():
    color  = COLORES.get(row["cocina"], "#888888")
    popup  = folium.Popup(hacer_popup(row), max_width=320)
    label  = f"#{row['id']}  {row['nombre']}"

    folium.CircleMarker(
        location     = [row["lat"], row["lon"]],
        radius       = 7,
        color        = color,
        fill         = True,
        fill_color   = color,
        fill_opacity = 0.85,
        weight       = 1.5,
        popup        = popup,
        tooltip      = label,
    ).add_to(contenedor)

restaurantes_layer.add_to(mapa)

# ─── 7. LEYENDA MANUAL ────────────────────────────────────────────────────────
leyenda_html = """
<div style="
    position: fixed;
    bottom: 30px;
    right: 15px;
    z-index: 1000;
    background: white;
    padding: 10px 14px;
    border-radius: 6px;
    border: 1px solid #ccc;
    font-family: Arial, sans-serif;
    font-size: 13px;
    box-shadow: 2px 2px 6px rgba(0,0,0,.15);">
  <b style="font-size:13px;">Tipo de cocina</b><br>
  <span style="color:#E63946;">&#9632;</span> Española<br>
  <span style="color:#457B9D;">&#9632;</span> Griega<br>
  <span style="color:#2A9D8F;">&#9632;</span> Italiana<br>
  <span style="color:#F4A261;">&#9632;</span> Mediterránea
</div>
"""
mapa.get_root().html.add_child(folium.Element(leyenda_html))

# ─── 8. CONTROL DE CAPAS ──────────────────────────────────────────────────────
folium.LayerControl(position="topright", collapsed=False).add_to(mapa)

# ─── 9. GUARDAR HTML ──────────────────────────────────────────────────────────
mapa.save(str(OUTPUT_HTML))

modo = "CLUSTERS" if USE_CLUSTERS else "PUNTOS INDIVIDUALES"
print(f"\n✅ Mapa guardado [{modo}]: {OUTPUT_HTML}")
print(f"   {len(df)} restaurantes | {df['cocina'].nunique()} tipos de cocina")
print(f"\n   Para cambiar entre modos, edita la línea:")
print(f"   USE_CLUSTERS = {USE_CLUSTERS}  →  cambia a {not USE_CLUSTERS}")
print(f"\n   Distribución final por barrio (top 10):")
print(df["barrio"].value_counts().head(10).to_string())

File: Scripts/02_mapa_restaurantes.R
# =============================================================
# MAPA INTERACTIVO - RESTAURANTES MEDITERRÁNEOS EN BOGOTÁ
# =============================================================
#
#   DoingEconomics_Proyect/         <- PROJECT_ROOT
#   ├── Data/
#   │   ├── Clean/
#   │   └── Raw/                    <- INPUT:  restaurantes_mediterraneos_bogota.csv
#   ├── Documentos/
#   ├── Outputs/
#   │   ├── Graphs/                 <- OUTPUT: mapa HTML + shapefile
#   │   └── Tables/
#   └── Scripts/                    <- este archivo vive aquí
#       ├── venv/
#       ├── .env
#       ├── 01_extraccion restaurantes.py
#       └── 02_mapa_restaurantes.R
#
# Paquetes requeridos:
#   install.packages(c("dplyr","sf","leaflet","leaflet.extras",
#                      "htmlwidgets","geodata","readr","glue"))
# =============================================================

library(dplyr)
library(sf)
library(leaflet)
library(leaflet.extras)
library(htmlwidgets)
library(readr)
library(glue)

# ── 0. RUTAS DEL PROYECTO ────────────────────────────────────────────────────
# Calcula rutas absolutas desde la ubicación de este script,
# sin depender de setwd() ni del usuario que lo ejecute.

SCRIPT_DIR   <- dirname(rstudioapi::getSourceEditorContext()$path)
PROJECT_ROOT <- dirname(SCRIPT_DIR)

# Carpetas del proyecto
DATA_RAW_DIR  <- file.path(PROJECT_ROOT, "Data",    "Raw")
DATA_CLEAN    <- file.path(PROJECT_ROOT, "Data",    "Clean")
GRAPHS_DIR    <- file.path(PROJECT_ROOT, "Outputs", "Graphs")
TABLES_DIR    <- file.path(PROJECT_ROOT, "Outputs", "Tables")
DOCS_DIR      <- file.path(PROJECT_ROOT, "Documentos")

# Crear carpetas si no existen
for (d in c(DATA_RAW_DIR, DATA_CLEAN, GRAPHS_DIR, TABLES_DIR)) {
  dir.create(d, recursive = TRUE, showWarnings = FALSE)
}

# Archivos de input y output
INPUT_CSV  <- file.path(DATA_RAW_DIR, "restaurantes_mediterraneos_bogota.csv")
OUTPUT_MAP <- file.path(GRAPHS_DIR,   "mapa_restaurantes_mediterraneos.html")
OUTPUT_SHP <- file.path(GRAPHS_DIR,   "restaurantes_mediterraneos_bogota.shp")

cat("── Rutas del proyecto ──────────────────────────────────\n")
cat(glue("  Script      : {SCRIPT_DIR}\n"))
cat(glue("  Raíz        : {PROJECT_ROOT}\n"))
cat(glue("  Input CSV   : {INPUT_CSV}\n"))
cat(glue("  Output HTML : {OUTPUT_MAP}\n"))
cat(glue("  Output SHP  : {OUTPUT_SHP}\n"))
cat("────────────────────────────────────────────────────────\n\n")

# ── 1. CARGAR Y PREPARAR DATOS DE RESTAURANTES ───────────────────────────────
df <- read_csv(INPUT_CSV, show_col_types = FALSE) |>
  filter(!is.na(lat), !is.na(lon)) |>
  mutate(
    id           = sprintf("%03d", row_number()),
    rating       = round(as.numeric(rating), 1),
    num_reviews  = as.integer(num_reviews),
    barrio       = ifelse(is.na(barrio) | barrio == "", "Sin datos", barrio),
    # Estrellas visuales para la tarjeta
    estrellas    = sapply(rating, function(r) {
      if (is.na(r)) return("Sin calificación")
      n <- round(r)
      paste0(strrep("★", n), strrep("☆", 5 - n), "  ", r, "/5")
    })
  )

sf_rest <- st_as_sf(df, coords = c("lon", "lat"), crs = 4326)

cat(glue("✅ {nrow(df)} restaurantes cargados\n\n"))

# ── 2. SHAPEFILE DE BOGOTÁ (Localidades) ──────────────────────────────────────
# Opción A: descarga automática con geodata (GADM, nivel 3 = municipios/localidades)
# Requiere: install.packages("geodata")
#
# Opción B (manual, más actualizado): descarga desde IDECA/DANE:
#   https://www.ideca.gov.co/recursos/mapas/localidad-bogota-dc
#   Guarda el shapefile en base_dir y usa: bogota_sf <- st_read("Loca.shp")

tryCatch({
  library(geodata)
  # Nivel 2 = departamentos, nivel 3 = municipios dentro de Cundinamarca
  colombia_mun <- gadm("COL", level = 3, path = tempdir()) |> st_as_sf()
  bogota_sf <- colombia_mun |>
    filter(grepl("Bogot", NAME_2, ignore.case = TRUE))

  if (nrow(bogota_sf) == 0) stop("No se encontró Bogotá en GADM nivel 3")
  cat("✅ Shapefile de Bogotá (GADM) cargado con", nrow(bogota_sf), "polígonos\n\n")

}, error = function(e) {
  # Fallback: shapefile simple de Bogotá desde GeoJSON público
  cat("⚠️  geodata no disponible, usando GeoJSON alternativo...\n")
  bogota_sf <<- st_read(
    "https://raw.githubusercontent.com/CodeforBogota/datasets/main/bogota-localidades.geojson",
    quiet = TRUE
  )
  cat("✅ GeoJSON de Bogotá cargado con", nrow(bogota_sf), "localidades\n\n")
})

# ── 3. PALETA DE COLORES POR TIPO DE COCINA ───────────────────────────────────
# Clasificar cada restaurante según su query de origen / tipos
df <- df |>
  mutate(
    cocina = case_when(
      grepl("grieg|greek",    tipos,        ignore.case = TRUE) ~ "Griega",
      grepl("grieg",          query_origen, ignore.case = TRUE) ~ "Griega",
      grepl("españ|spanish",  tipos,        ignore.case = TRUE) ~ "Española",
      grepl("españ|tapas",    query_origen, ignore.case = TRUE) ~ "Española",
      grepl("italian",        tipos,        ignore.case = TRUE) ~ "Italiana",
      grepl("italian|pizza|trattoria", query_origen, ignore.case = TRUE) ~ "Italiana",
      TRUE ~ "Mediterránea"
    )
  )

sf_rest$cocina <- df$cocina

pal_cocina <- colorFactor(
  palette = c("#E63946", "#457B9D", "#2A9D8F", "#F4A261"),
  domain  = c("Griega", "Española", "Italiana", "Mediterránea")
)

# ── 4. CONSTRUIR POPUP HTML PARA CADA RESTAURANTE ─────────────────────────────
make_popup <- function(id, nombre, direccion, barrio, rating,
                       num_reviews, estrellas, cocina, google_maps_url) {
  reviews_txt <- ifelse(is.na(num_reviews), "Sin datos",
                        format(num_reviews, big.mark = "."))
  maps_link   <- ifelse(is.na(google_maps_url) | google_maps_url == "",
                        "",
                        glue('<a href="{google_maps_url}" target="_blank">
                               Ver en Google Maps ↗</a>'))
  glue('
  <div style="
    font-family: Arial, sans-serif;
    min-width: 220px;
    max-width: 280px;
    padding: 4px;">

    <div style="
      background: #1d3557;
      color: white;
      padding: 8px 12px;
      border-radius: 6px 6px 0 0;
      margin: -4px -4px 8px -4px;">
      <span style="font-size:11px; opacity:.8;">#{id}</span>
      <div style="font-size:15px; font-weight:bold; margin-top:2px;">{nombre}</div>
      <span style="
        background: {pal_cocina(cocina)};
        font-size:10px;
        padding: 2px 6px;
        border-radius: 10px;
        margin-top: 4px;
        display: inline-block;">
        {cocina}
      </span>
    </div>

    <table style="width:100%; font-size:12px; border-collapse:collapse;">
      <tr>
        <td style="color:#555; padding:3px 0; width:80px;">📍 Barrio</td>
        <td style="font-weight:500;">{barrio}</td>
      </tr>
      <tr>
        <td style="color:#555; padding:3px 0;">🏠 Dirección</td>
        <td style="font-size:11px;">{direccion}</td>
      </tr>
      <tr>
        <td style="color:#555; padding:3px 0;">⭐ Rating</td>
        <td>{estrellas}</td>
      </tr>
      <tr>
        <td style="color:#555; padding:3px 0;">💬 Reseñas</td>
        <td>{reviews_txt}</td>
      </tr>
    </table>

    <div style="margin-top:8px; font-size:11px; color:#457b9d;">
      {maps_link}
    </div>
  </div>
  ')
}

popups <- mapply(make_popup,
  id             = df$id,
  nombre         = df$nombre,
  direccion      = df$direccion,
  barrio         = df$barrio,
  rating         = df$rating,
  num_reviews    = df$num_reviews,
  estrellas      = df$estrellas,
  cocina         = df$cocina,
  google_maps_url = df$google_maps_url,
  SIMPLIFY       = TRUE
)

# ── 5. CONSTRUIR MAPA LEAFLET ─────────────────────────────────────────────────
mapa <- leaflet() |>

  # Tiles base
  addProviderTiles(providers$CartoDB.Positron, group = "Claro") |>
  addProviderTiles(providers$Esri.WorldStreetMap, group = "Detallado") |>

  # Shapefile de Bogotá (localidades)
  addPolygons(
    data        = bogota_sf,
    fillColor   = "#f0f0f0",
    fillOpacity = 0.25,
    color       = "#888888",
    weight      = 1,
    opacity     = 0.7,
    group       = "Localidades",
    label       = ~if ("NAME_3" %in% names(bogota_sf)) NAME_3
                   else if ("LocNombre" %in% names(bogota_sf)) LocNombre
                   else NULL
  ) |>

  # Puntos de restaurantes
  addCircleMarkers(
    data         = sf_rest,
    radius       = 7,
    color        = ~pal_cocina(cocina),
    fillColor    = ~pal_cocina(cocina),
    fillOpacity  = 0.85,
    weight       = 1.5,
    stroke       = TRUE,
    popup        = popups,
    label        = ~glue("#{id} {nombre}"),
    labelOptions = labelOptions(
      style     = list("font-weight" = "normal", padding = "3px 6px"),
      textsize  = "13px",
      direction = "auto"
    ),
    group        = "Restaurantes",
    clusterOptions = markerClusterOptions(
      showCoverageOnHover = FALSE,
      zoomToBoundsOnClick = TRUE,
      spiderfyOnMaxZoom   = TRUE
    )
  ) |>

  # Leyenda por tipo de cocina
  addLegend(
    position = "bottomright",
    pal      = pal_cocina,
    values   = df$cocina,
    title    = "Tipo de cocina",
    opacity  = 0.9
  ) |>

  # Control de capas
  addLayersControl(
    baseGroups    = c("Claro", "Detallado"),
    overlayGroups = c("Localidades", "Restaurantes"),
    options       = layersControlOptions(collapsed = FALSE)
  ) |>

  # Buscador por nombre
  addSearchFeatures(
    targetGroups = "Restaurantes",
    options      = searchFeaturesOptions(
      propertyName = "label",
      zoom         = 16,
      openPopup    = TRUE,
      hideMarkerOnCollapse = TRUE
    )
  ) |>

  # Vista inicial centrada en Bogotá
  setView(lng = -74.0721, lat = 4.7110, zoom = 11)

# ── 6. GUARDAR HTML ───────────────────────────────────────────────────────────
saveWidget(mapa, file = OUTPUT_MAP, selfcontained = TRUE)
cat(glue("\n✅ Mapa guardado: {OUTPUT_MAP}\n"))
cat(glue("   {nrow(df)} restaurantes | {length(unique(df$cocina))} tipos de cocina\n"))

# ── 7. EXPORTAR SHAPEFILE ─────────────────────────────────────────────────────
st_write(sf_rest, OUTPUT_SHP, delete_dsn = TRUE, quiet = TRUE)
cat(glue("✅ Shapefile guardado: {OUTPUT_SHP}\n"))